### ISSUE TITLE
Contextual Dependency Clarification

### ISSUE THREAD

**Issue by n0nag0n** [mako-framework/docs#29]
Hey there,

Been thinking about how you guys do dependency injection and in your docs on this page [https://makoframework.com/docs/9.0/getting-started:dependency-injection](https://makoframework.com/docs/9.0/getting-started:dependency-injection) you mention contextual injection. It feels like it's important to know, but I don't think I quite understand what the goal of it is.
![image](https://user-images.githubusercontent.com/2322095/219478753-17a1b244-f458-4448-960e-707508cde0d3.png)
Is there a real life example that could better explain how it works (for me and others)?

**Comment by freost** [comment]
It's probably the least used feature of the container but it was useful for us when we had two versions on an API while we migrated from Sphinx to Elastic. Both the endpoint controllers required an instance of SearchInterface. One got a SphinxSearch instance while the other got the ElasticSearch instance. It allowed us to run both versions in parallel until Sphinx was replaced.

**Comment by n0nag0n** [comment]
So if I'm understanding your example correctly, would you run something like this:
```php
$container->registerContextualDependency(Api_v1_Controller::class, SearchInterface::class, SphinxSearch::class);
$container->registerContextualDependency(Api_v2_Controller::class, SearchInterface::class, ElasticSearch::class);
```

**Comment by freost** [comment]
Yeah that's right. It's probably not something you'll use often but it was nice to have when we needed it.
